dev/core#3542 Store extdir cache in sqlcache rather than the filesystem

mlutfy opened the pull request:
Overview
----------------------------------------

The in-app Extension Browser currently has an ad-hoc cache on the filesystem. This PR uses an SQL cache instead. It reduces the complexity/risks of filesystem permissions, at least to get started, because the cache directory may not be writable.

Related to [dev/cloud-native#17](https://lab.civicrm.org/dev/cloud-native/-/issues/17)  and PR #12624.

Before
----------------------------------------

Filesystem cache.

After
----------------------------------------

Standard SQL cache.

Technical Details
----------------------------------------

Introduces a new container cache `extension_browser`. This was based on #12624. I guess it makes sense so that we can allow more flexibility in storing that specific cache somewhere else, but I have to admit those container names seem kind of arbitrary.

I also hesitate to remove `checkRequirements`, which is now empty, because maybe we'll think of other things to check, but if not, I guess may as well remove it?

This PR also might change the cache TTL. By default sqlcache has a 6h TTL. This seems fine to me, considering there is a big "refresh" button on the browser.

mlutfy committed:
dev/cloud-native#17 Store extdir cache in sqlcache rather than the filesystem

mattwire commented:
@bgm conflicts and style issues

mlutfy commented:
Thanks - fixed.

totten commented on CRM/Extension/Browser.php:
```diff
@@ -221,12 +177,8 @@ private function _discoverRemote() {
-      $json = file_get_contents($filename);
-    }
-    if (empty($json)) {
+    $json = \Civi::cache('extension_browser')->get('extdir');
```
Fun fact: There can be multiple URLs, eg

* The sysadmin can configure the `ext_repo_url` setting (eg to enable alpha/beta downloads).
* The sysadmin can use CLI options in `cv dl` (like `--dev`, `--filter-ver`, `--filter-status`) which grab from another feed.

As in the `$filename` above, the cache-key (`get($key)` and `set($key,$value)`) should include the URL. Something like:

```php
$json = \Civi::cache('extension_browser')->get('extdir_' . $this->getRepositoryUrl());
$json = \Civi::cache('extension_browser')->get('extdir_' . md5($this->getRepositoryUrl()));
$json = \Civi::cache('extension_browser')->get(md5($this->getRepositoryUrl()));
```

totten commented:
> This PR also might change the cache TTL. By default sqlcache has a 6h TTL. This seems fine to me, considering there is a big "refresh" button on the browser.

+1

> I also hesitate to remove checkRequirements, which is now empty, because maybe we'll think of other things to check, but if not, I guess may as well remove it?

Yeah, I think it's OK either way. You could add a comment like `// If this hasn't found new use, remove after 5.60`.

> [Test Result](https://test.civicrm.org/job/CiviCRM-Core-PR/49703/testReport/) (2 failures / +2)
> * [CRM_Extension_BrowserTest.testCheckRequirements_BadCachedir_false](https://test.civicrm.org/job/CiviCRM-Core-PR/49703/testReport/junit/(root)/CRM_Extension_BrowserTest/testCheckRequirements_BadCachedir_false/)
> * [CRM_Extension_BrowserTest.testCheckRequirements_BadCachedir_nonexistent](https://test.civicrm.org/job/CiviCRM-Core-PR/49703/testReport/junit/(root)/CRM_Extension_BrowserTest/testCheckRequirements_BadCachedir_nonexistent/)

The old requirements have some narrow/corresponding test-cases that should be removed.

> Introduces a new container cache `extension_browser`. This was based on https://github.com/civicrm/civicrm-core/pull/12624. I guess it makes sense so that we can allow more flexibility in storing that specific cache somewhere else, but I have to admit those container names seem kind of arbitrary.

Yeah, they are a bit arbitrary. I think it's fine as it is. If you wanted something that feels less bespoke:

* There's `Civi::cache('long')` which should also be suitable (https://docs.civicrm.org/dev/en/latest/framework/cache/).
* A general `http` cache could be useful. There is [a Guzzle plugin](https://github.com/Kevinrob/guzzle-cache-middleware) with a [PSR-16 adapter](https://github.com/Kevinrob/guzzle-cache-middleware/blob/master/src/Storage/Psr16CacheStorage.php). It could potentially serve some other use-cases, and (IIRC) it respects the HTTP cache/ttl headers. However, that would be a bigger patch.

mlutfy commented:
@totten Thanks for the detailed review! I updated the PR:

- Fixed the `cacheKey`, using an md5 of the URL as you suggested.
- Updated the tests, to remove the 2 deprecated tests, and also updated the constructor calls in some other tests (because the 3rd argument was removed)

I kept the `extension_browser` container, because it makes it more convenient to `flush()`.

totten commented:
Looks good @mlutfy.

I did some `r-run` with web UI, `cv ext:list`, and `mysql` CLI (to inspect/hack values). It seemed to work pretty well. :+1: 

Another thought, though: The current cache-definition allows flexible storage backends (Redis, Memcache, MySQL), and they have different storage limits.

* The approved/stable JSON feed is 150-200kb. The unapproved/dev feed is 500-700kb. (Size varies by encoding - eg some caches store base64-encoded data, which uses more space.)
* MySQL's default [max_allowed_packet](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_max_allowed_packet) is 64mb. The `civicrm_cache.data` is `longtext` (*ie massive*).
* Redis strings are limited to [512 mb](https://redis.io/docs/manual/data-types/)
* Memcached's [default limit is 1mb](https://github.com/memcached/memcached/wiki/ReleaseNotes142#configurable-maximum-item-size).

This means that we're currently within bounds, but there isn't much extra space between the largest feed (unapproved exts, base64, 700kb) and the smallest storage (memcache, 1mb). But we're well within bounds for the typical (approved/stable) feed.

I'm not sure it really needs to be in Memcache. (It's not read frequently.) It would be OK to control the `type` of storage more tightly.

```php
diff --git a/Civi/Core/Container.php b/Civi/Core/Container.php
index 207ab82268..8dfa653c58 100644
--- a/Civi/Core/Container.php
+++ b/Civi/Core/Container.php
@@ -167,7 +167,6 @@ class Container {
       'fields' => 'contact fields',
       'contactTypes' => 'contactTypes',
       'metadata' => 'metadata',
-      'extension_browser' => 'extension_browser',
     ];
     $verSuffixCaches = ['metadata'];
     $verSuffix = '_' . preg_replace(';[^0-9a-z_];', '_', \CRM_Utils_System::version());
@@ -189,6 +188,16 @@ class Container {
       ))->setFactory('CRM_Utils_Cache::create')->setPublic(TRUE);
     }
 
+    $container->setDefinition('cache.extension_browser', new Definition(
+      'CRM_Utils_Cache_Interface',
+      [
+        [
+          'name' => 'extension_browser',
+          'type' => ['SqlGroup', 'ArrayCache'],
+        ],
+      ]
+    ))->setFactory('CRM_Utils_Cache::create')->setPublic(TRUE);
+
     // PrevNextCache cannot use memory or array cache at the moment because the
     // Code in CRM_Core_BAO_PrevNextCache assumes that this cache is sql backed.
     $container->setDefinition("cache.prevNextCache", new Definition(
```

I'll give it a `merge ready` -- if you strongly prefer the current/flexible-storage, that's OK. But you should consider switching to SQL.

mlutfy commented:
Makes sense - I amended the commit. Thanks!

totten commented:
Looks good. Merging.
